scipy/stats/stats.py

```diff
@@ -4363,11 +4360,18 @@ def ks_2samp(data1, data2):
     data2 = np.sort(data2)
     n1 = data1.shape[0]
     n2 = data2.shape[0]
-    data_all = np.concatenate([data1, data2])
-    cdf1 = np.searchsorted(data1, data_all, side='right') / (1.0*n1)
-    cdf2 = np.searchsorted(data2, data_all, side='right') / (1.0*n2)
-    d = np.max(np.absolute(cdf1 - cdf2))
-    # Note: d absolute not signed distance
+    common_type = np.find_common_type([], [data1.dtype, data2.dtype])
+    if not (np.issubdtype(common_type, np.number) and
+            not np.issubdtype(common_type, np.complexfloating)):
+        raise ValueError('ks_2samp only accepts real inputs')
+    if np.any(np.isnan(data1)) or np.any(np.isnan(data2)):
+        raise ValueError('ks_2samp only accepts non-nan inputs')
```

Comment:
an optimization to avoid the overhead for the standard case of no nans

np.sort moves nans at the end, so only data1[-1] and data2[-2] need to be checked for isnan